daemon/cmd/endpoint.go

```diff
@@ -305,6 +298,21 @@ func (d *Daemon) createEndpoint(ctx context.Context, epTemplate *models.Endpoint
 		return d.errorDuringCreation(ep, fmt.Errorf("unable to insert endpoint into manager: %s", err))
 	}
 
+	// We need to update the the visibility policy after adding the endpoint in
+	// the endpoint manager because the endpoint manager create the endpoint
+	// queue of the endpoint. If we execute this function before the endpoint
+	// manager creates the endpoint queue the operation will fail.
+	if ep.K8sNamespaceAndPodNameIsSet() && k8s.IsEnabled() && k8sLabelsConfigured {
+		ep.UpdateVisibilityPolicy(func(ns, podName string) (proxyVisibility string, err error) {
```

Comment:
Doesn't this force two policy recalculations for all relevant endpoints?

`ep.UpdateVisibilityPolicy()` originally cached the regeneration flag and configure the labels to ensure that the initial regeneration takes into account the visibility policy. The queueing ensured ordering for handling of events, but introduced this problem where the queue wasn't yet initialized. Clearly we need to fix that. But then with this PR, because the endpoint was exposed it's possible for other events to enqueue and the initial regeneration to kick off before this pod annotation is taken into account.

Could/should we maybe force this event to be pushed onto the eventqueue during endpoint eventqueue creation to ensure it happens first to reduce the number of policy recalculations?

Or are you thinking it's better to fix the immediate bug here and now, and worry about optimizing this later?

Reply:
The answer is no, there's likely no additional regeneration. This is still occurring before `ep.UpdateLabels()` and `ep.RegenerateAfterCreation()` so the only way would be if some other cluster event came through and found the endpoint via endpointManager and triggered from there. This is both unlikely, and also would otherwise cause two regenerations anyway.